### PR TITLE
Fixed visibility issue in split-brain recovery

### PR DESCRIPTION
**Pull request by Donnerbart**
* added `AbstractContainerCollector` which collects non-threadsafe data structure containers via `PartitionSpecificRunnables`
* added `AbstractContainerMerger` which merges collected data containers
* added data structure specific merge helpers for `IQueue`, `ISet`, `IList`, `MultiMap`, `IAtomicLong`, `IAtomicReference`, `Ringbuffer`, `CardinalityEstimator` and `ScheduledExecutorService`
* added `clear()` method to `RingbufferContainer` and `Ringbuffer`
* added a minimal timeout of 5 seconds to wait for merge operations
* reduced memory footprint of `AtomicLong` and `AtomicReference` containers by removing name, config and `SerializationService` reference, which were formerly added by split-brain healing

<hr>

New classes in `com.hazelcast.spi.impl.merge`:
screenshot from 2018-03-07 14-04-01

Class hierarchy:
container-collector

The `AbstractNonPartitionedContainerCollector` has two groups of implementations:
* `CollectionContainerCollector` and `QueueContainerCollector`
* `AtomicLongContainerCollector`, `AtomicReferenceContainerCollector` and `CardinalityEstimatorContainerCollector`

The first group is fine as it is. For the second one I could add another abstract class to de-duplicate the lookup maps. But I think the small amount of duplicated code doesn't justify that additional layer of abstraction.

<hr>

The `ReplicatedMap` will be fixed in another PR, since it will be aligned with `IMap` and `ICache`, not the container based data structures of this PR.

<hr>

There is also no cleanup of any container maps in the `prepareMergeRunnable()` methods anymore. Just a simple:
```java
@Override
public Runnable prepareMergeRunnable() {
    MyContainerCollector collector = new MyContainerCollector(nodeEngine, containers);
    collector.run();
    return new Merger(collector);
}
```
**Note:** The collector will retain the collected containers for the merge runnable. So `ManagedService#reset()` is not allowed to clear or destroy those containers. To ensure this the collector removes the link between the data structure and a container via `Iterator#remove()` in `AbstractContainerCollector$CollectContainerRunnable.run()`.

The cleanup of the containers itself is done in `AbstractContainerCollector.destroy()` via `AbstractContainerCollector.destroy(container)` after the merge is done, or directly if the container is not collected. Containers from backup partitions are directly cleaned via `AbstractContainerCollector.destroyBackup(container)`.

This info has also been added to the JavaDoc of `AbstractContainerCollector`.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+ * The cleanup of the containers itself is done via {@link #destroy(Object)} after the merge is done
+ * or if the container is not collected. Containers from backup partitions are directly cleaned via
+ * {@link #destroyBackup(Object)}.
+ *
+ * @param <C> container of the data structure
+ */
+@SuppressWarnings("WeakerAccess")
+public abstract class AbstractContainerCollector<C> {
+
+    private final ConcurrentMap<Integer, Collection<C>> containersByPartitionId
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: This `ConcurrentMap` ensures the visibility between the threads. It's populated by `PartitionSpecificRunnable` on the partition threads and read from by the `AbstractContainerMerger` on an arbitrary thread.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+                }
+            } else {
+                destroyBackup(container);
+            }
+        }
+
+        private Collection<C> getContainerCollection() {
+            Collection<C> containers = containersByPartitionId.get(partitionId);
+            if (containers == null) {
+                Collection<C> candidate = new ConcurrentLinkedQueue<C>();
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: We concurrently start one `CollectNonPartitionedContainerRunnable` per non-partitioned container in the data structure. The containers can all be in the same partition or in different ones. Since the final map will be a `ConcurrentMap<Integer, Collection<Container>>`, we also need a concurrent collection here. Otherwise the concurrent `Runnable`s can't populate it in parallel.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        if (size == 0) {
+            return;
+        }
+
+        runInternal();
+
+        assert operationCount > 0 : "No merge operations have been invoked in AbstractContainerMerger";
+
+        try {
+            long timeoutMillis = Math.max(size * TIMEOUT_FACTOR, MINIMAL_TIMEOUT_MILLIS);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java]: I've added this, so we wait at least 5 seconds for all merge operations to finish (even for a single one). If the operations finish faster, we won't wait any longer. I just wanted to prevent an eagerly sent `MERGED` lifecycle event. Maybe we should even define a longer minimum timeout value?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+            // collect partitioned containers
+            int partitionCount = partitionService.getPartitionCount();
+            latch = new CountDownLatch(partitionCount);
+
+            for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+                operationExecutor.execute(new CollectPartitionedContainerRunnable(partitionId));
+            }
+        } else {
+            // collect non-partitioned containers
+            Collection<String> containerNames = containerNames();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: Maybe I'm wrong but it seems there might be a race here. We first collect the containers and then collect their data on the partition thread. In the meantime, a new container might get created or an existing one might get destroyed (if the partition operation was already being executed while the node state was reset).
We should probably just run partition operations, regardless if it's a partitioned container or not, and collect both the names and data on the partition thread.
The downside being that once on the partition thread, you need to collect only the containers that belong to that partition.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: Hmm, you might be right, yeah. This will also get rid of the concurrent list, which is just needed since there might be multiple runnables per partition. I'll give it a try!

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: That was really a good one, the code is even simpler now :)

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        if (valueCount == 0) {
+            return;
+        }
+
+        runInternal();
+
+        assert operationCount > 0 : "No merge operations have been invoked in AbstractContainerMerger";
+
+        try {
+            long timeoutMillis = Math.max(valueCount * TIMEOUT_FACTOR, MINIMAL_TIMEOUT_MILLIS);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java]: I've added this, so we wait at least 5 seconds for all merge operations to finish (even for a single one). If the operations finish faster, we won't wait any longer. I just wanted to prevent an eagerly sent MERGED lifecycle event. Maybe we should even define a longer minimum timeout value?

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java [hunk 6]
```diff
@@ -347,126 +337,56 @@ public void ensureQuorumPresent(String distributedObjectName, QuorumType require
...
-            }
-        }
-
-        // clear all items either owned or backup
-        reset();
-
-        return new Merger(partitionContainerMap);
+        RingbufferContainerCollector collector = new RingbufferContainerCollector(nodeEngine, containers);
+        collector.run();
+        return new Merger(collector);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java]: Same as in `MultiMapService`, the containers are used after they have been `reset()` but no harm here since only the collection `containers` is cleared.

File: hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/DistributedScheduledExecutorService.java [hunk 7]
```diff
@@ -332,116 +310,54 @@ public String getQuorumName(final String name) {
...
+                Collection<ScheduledExecutorContainer> containers = entry.getValue();
+
+                for (ScheduledExecutorContainer container : containers) {
+                    String name = container.getName();
+                    MergePolicyConfig mergePolicyConfig = collector.getMergePolicyConfig(container);
+                    SplitBrainMergePolicy mergePolicy = getMergePolicy(mergePolicyConfig);
+                    int batchSize = mergePolicyConfig.getBatchSize();
+
+                    mergingEntries = new ArrayList<MergingEntryHolder<String, ScheduledTaskDescriptor>>(batchSize);
+                    for (ScheduledTaskDescriptor descriptor : container.prepareForReplication(true).values()) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/DistributedScheduledExecutorService.java]: Now this is called after the call to `ScheduledExecutorService.reset()` but I think it works ok.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java [hunk 8]
```diff
@@ -502,136 +492,60 @@ public void ensureQuorumPresent(String distributedObjectName, QuorumType require
...
-                    map.put(name, container);
-                }
-            }
-
-            // clear all items either owned or backup
-            partitionContainer.containerMap.clear();
-        }
-
-        return new Merger(itemMap);
+        MultiMapContainerCollector collector = new MultiMapContainerCollector(nodeEngine, partitionContainers);
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java]: Ignore this, now I see how it works

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+    /**
+     * Returns the partition ID of a non-partitioned container of the data structure.
+     * <p>
+     * Will be used if {@link #isPartitioned()} is {@code false}.
+     * <p>
+     * Can be overridden if a custom partition ID calculation is needed.
+     *
+     * @return the partition ID of a container
+     */
+    protected int getContainerPartitionId(String containerName) {
```
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: See last line of JavaDoc. Can be overridden if there is another way.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            if (!semaphore.tryAcquire(operationCount, timeoutMillis, TimeUnit.MILLISECONDS)) {
+                logger.warning("Split-brain healing for " + getLabel() + " didn't finish within the timeout...");
+            }
+        } catch (InterruptedException e) {
+            logger.finest("Interrupted while waiting for split-brain healing of " + getLabel() + "...");
+            Thread.currentThread().interrupt();
+        } finally {
+            collector.destroy();
```
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java]: Is the operation serialized right away after the operation is submitted to the executor? I remember some kind of test failures in one of the many implementation, when I cleared the snapshot data immediately. But it could very well be the `OperationFactory` and `IMap`, not one of these data structures.

Since the node is a lite member until the merge operations are completed, it should not hurt to occupy the memory a tiny bit longer, so the `collector.destroy()` in the `finally` block just feels a bit safer for me.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java]: I moved to the inner class, since it's unlikely that we need an override any time soon.

File: hazelcast/src/main/java/com/hazelcast/cardinality/impl/CardinalityEstimatorContainerCollector.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        this.containerNames = createHashMap(containers.size());
+        this.containerPolicies = createHashMap(containers.size());
+        initLookupMaps(nodeEngine, containers);
+    }
+
+    /**
+     * The {@link CardinalityEstimatorContainer} doesn't know its name or configuration, so we create these lookup maps.
+     * This is cheaper than storing this information permanently in the container.
+     */
+    private void initLookupMaps(NodeEngine nodeEngine, Map<String, CardinalityEstimatorContainer> containers) {
```
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/cardinality/impl/CardinalityEstimatorContainerCollector.java]: Good one! I've added an `onRemoveContainer(String containerName, C container)` hook in `AbstractNonPartitionedContainerCollector` (no need for this for partitioned ones). This is now invoked on the partition threads.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractPartitionedContainerCollector.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import com.hazelcast.spi.NodeEngine;
+
+import java.util.Collection;
+
+/**
+ * Base implementation of {@link AbstractContainerCollector} for partitioned data structures.
+ *
+ * @param <C> container of the data structure
+ */
+public abstract class AbstractPartitionedContainerCollector<C> extends AbstractContainerCollector<C> {
```
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractPartitionedContainerCollector.java]: Yeah, the naming of the collectors is misleading. We should change this. Maybe something like `AbstractPartitionContainerCollector` and `AbstractNamedContainerCollector` fits better. And internally use `hasIterator()` instead of `isPartitioned()` to distinct (which also gives a better hint which methods are used).

I'll also have a look if we can use iterators for the named containers.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java]: > Is the operation serialized right away after the operation is submitted to the executor?

Operations are [serialized just before being sent out over the network](https://github.com/hazelcast/hazelcast/blob/00ee8406b7945b601fc6c2e3783b7c53a5dd582d/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OutboundOperationHandler.java#L51) and it will be serialized again on each retry if it fails with retryable exception, so it's not safe to destroy the collector before operations are done.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractNonPartitionedContainerCollector.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Map;
+
+/**
+ * Base implementation of {@link AbstractContainerCollector} for non-partitioned data structures.
+ *
+ * @param <C> container of the data structure
+ */
+public abstract class AbstractNonPartitionedContainerCollector<C> extends AbstractContainerCollector<C> {
+
+    protected final Map<String, C> containers;
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractNonPartitionedContainerCollector.java]: We should probably enforce this to be `ConcurrentMap` since we know it will be accessed from different threads.

File: hazelcast/src/main/java/com/hazelcast/concurrent/atomiclong/AtomicLongContainerCollector.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+class AtomicLongContainerCollector extends AbstractNonPartitionedContainerCollector<AtomicLongContainer> {
+
+    private final Config config;
+    private final Map<AtomicLongContainer, String> containerNames;
+    private final Map<AtomicLongContainer, MergePolicyConfig> containerPolicies;
+
+    AtomicLongContainerCollector(NodeEngine nodeEngine, Map<String, AtomicLongContainer> containers) {
+        super(nodeEngine, containers);
+        this.config = nodeEngine.getConfig();
+        this.containerNames = createHashMap(containers.size());
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/atomiclong/AtomicLongContainerCollector.java]: These will have to be concurrent maps since they are accessed concurrently from different partition threads. Applies to other collectors which have them as well.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractPartitionedContainerCollector.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractPartitionedContainerCollector.java]: I've implemented the iterator for the `AbstractNamedContainerCollector` (needed a little tweak, I think you version never made progress again since `advance()` was just called once). So the `AbstractPartitionContainerCollector` could be completely removed (together with a bunch of methods).

File: hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java [hunk 15]
```diff
@@ -390,130 +383,62 @@ public String getQuorumName(String name) {
...
+                int partitionId = entry.getKey();
+                Collection<QueueContainer> containerList = entry.getValue();
+                for (QueueContainer container : containerList) {
+                    Deque<QueueItem> itemList = container.getItemQueue();
 
-                    int batchSize = container.getConfig().getMergePolicyConfig().getBatchSize();
-                    SplitBrainMergePolicy mergePolicy = getMergePolicy(container);
                     String name = container.getName();
+                    int batchSize = container.getConfig().getMergePolicyConfig().getBatchSize();
+                    SplitBrainMergePolicy mergePolicy = getMergePolicy(container.getConfig().getMergePolicyConfig());
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: Minor: you can use `((QueueContainerCollector) collector).getMergePolicyConfig(container)` here as well. Same in `CollectionService` and `MultiMapService`, possibly others as well.

Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java]: Those containers know their configuration, that workaround via the collector was just meant for the lean containers of `AtomicLong`, `AtomicReference` and `CardinalityEstimator`, which cannot provide that info on their own.

I also think the code is easier to read without the casting.
```java
mergePolicy = getMergePolicy(container.getConfig().getMergePolicyConfig());
```
vs.
```java
mergePolicy = getMergePolicy(((QueueContainerCollector) collector).getMergePolicyConfig(container));
```
The collector would anyway call `container.getConfig().getMergePolicyConfig()` again. I guess we can save HotSpot a tiny bit of work here.

**Comment by Donnerbart** [comment]
`ConcurrentMap` is now enforced in all collectors. The name based ones also provide an `Iterator` now, so the code is much cleaner now (just a single collect mechanism). The new `EmptyIterator` and `ContainerIterator` are also tested.

File: hazelcast/src/test/java/com/hazelcast/spi/impl/merge/AbstractContainerMergerTest.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+        emptyCollector = new TestContainerCollector(nodeEngine, false, true);
+        emptyCollector.run();
+    }
+
+    /**
+     * Tests that the merger finished under normal conditions.
+     */
+    @Test
+    @RequireAssertEnabled
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/spi/impl/merge/AbstractContainerMergerTest.java]: Why is this necessary? Also, in order to be taken into account, it seems a `@Rule` is missing:
```
 * To have this annotation honored, a test class must define an instance of {@link AssertEnabledFilterRule}.
 * This is automatically done via {@link AutoRegisteredTestRule}.
```

Comment (Donnerbart) [reply hazelcast/src/test/java/com/hazelcast/spi/impl/merge/AbstractContainerMergerTest.java]: It's a safety net to ensure that asserts are enabled (it will let the test fail if they are not working). It's also a nice marker for a test, that it tests an `assert` in the production code, so most probably here will be no `assertEquals()` or so in the test code.

The rule is automatically added via `AutoRegisteredTestRule`. The JavaDoc maybe a little misleading here.
